src/app/clusters/ota-requestor/BDXDownloader.cpp

```diff
@@ -35,6 +36,51 @@ using chip::bdx::TransferSession;
 
 namespace chip {
 
+void TransferTimeoutCheckHandler(System::Layer * systemLayer, void * appState)
+{
+    VerifyOrReturn(appState != nullptr);
+    BDXDownloader * bdxDownloader = static_cast<BDXDownloader *>(appState);
+
+    if (bdxDownloader->CheckTransferTimeout())
+    {
+        // End download if transfer timeout has been detected
+        bdxDownloader->OnDownloadTimeout();
+    }
+    else
+    {
+        // Else restart the timer
+        systemLayer->StartTimer(bdxDownloader->GetTimeout(), TransferTimeoutCheckHandler, appState);
+    }
+}
+
+System::Clock::Timeout BDXDownloader::GetTimeout()
+{
+    return mTimeout;
+}
+
+void BDXDownloader::Reset()
+{
+    mPrevPercentageComplete = 0;
+    DeviceLayer::SystemLayer().StartTimer(mTimeout, TransferTimeoutCheckHandler, this);
+}
+
+bool BDXDownloader::CheckTransferTimeout()
+{
+    uint8_t curPercentageComplete =
+        mImageProcessor->GetPercentComplete().IsNull() ? 0 : mImageProcessor->GetPercentComplete().Value();
+
+    if (curPercentageComplete > mPrevPercentageComplete)
```

Comment:
Instead of percentage, using the block counter would be better. Each block transfer increases the block counter, whereas integer percentage of a large image may not change fast (or enough) over 5 minutes.